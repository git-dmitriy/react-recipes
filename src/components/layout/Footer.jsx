import GithubIcon from '../icons/GithubIcon';
import { FaGithub } from 'react-icons/fa';

export function Footer() {
  return (
    <footer className='bg-gray-900 py-3 text-white'>
      <div className='container mx-auto'>
        <div className='flex justify-center items-center'>
          Â© {new Date().getFullYear()} Dmitriy Shalberkin
          <a
            className='mx-4 w-10 text-4xl hover:text-gray-400 transition duration-300'
            target='_blank'
            rel='noreferrer'
            href='https://github.com/git-dmitriy'
          >
            <GithubIcon />
          </a>
        </div>
      </div>
    </footer>
  );
}
