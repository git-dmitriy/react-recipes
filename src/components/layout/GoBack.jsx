import { useEffect, useState } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import { FaArrowLeft } from 'react-icons/fa';

const GoBack = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const [isHome, setIsHome] = useState(false);

  useEffect(() => {
    if (location.pathname === '/') {
      setIsHome(true);
    } else {
      setIsHome(false);
    }
  }, [location]);

  return (
    <div className='flex items-center sm:mx-6'>
      {!isHome ? (
        <button className='w-5 text-3xl' onClick={() => navigate(-1)}>
          <FaArrowLeft className='fill-current' />
        </button>
      ) : (
        <div className='w-5 text-transparent'>
          <FaArrowLeft className='fill-current text-transparent' />
        </div>
      )}
    </div>
  );
};

export default GoBack;
