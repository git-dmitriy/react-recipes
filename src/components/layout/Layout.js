export const Layout = ({ children }) => {
  return (
    <main className='text-gray-600'>
      <div className='container px-5 pt-2 pb-24 mx-auto flex flex-col'>
        <div className='w-full lg:w-6/6 2xl:w-4/6 mx-auto '>{children}</div>
      </div>
    </main>
  );
};
