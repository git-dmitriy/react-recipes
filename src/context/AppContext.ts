import { createContext } from 'react';
import { ContextTypes } from '@/appTypes';

export const AppContext = createContext({} as ContextTypes);
