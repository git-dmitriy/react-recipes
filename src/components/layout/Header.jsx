import { useNavigate, NavLink } from 'react-router-dom';
import { Search } from '../Search/Search';
import { GoBack } from './GoBack';
import { BsFillBookmarkFill } from 'react-icons/bs';

export const Header = () => {
  const { push } = useNavigate();

  const handleSearch = (searchQuery) => {
    if (searchQuery.trim().length !== 0) {
      push({
        pathname: '/recipes',
        search: `search=${searchQuery}`,
      });
    }
  };

  const textLinkClassName =
    'border-b-2 border-transparent font-bold outline-none hover:border-red-500 mx-1.5 sm:mx-6 focus:outline-none focus:border-red-500 transition';
  const iconLinkClassName =
    'border-b-2 border-transparent outline-none hover:text-red-500 mx-1.5 sm:mx-6 focus:outline-none focus:text-red-500 transition text-2xl';
  const activeTextLink = ' border-b-2 border-red-500';
  const activeIconLink = ' text-red-500';

  return (
    <nav className='bg-yellow-300 shadow dark:bg-gray-800 fixed top-0 w-full backdrop-blur-md backdrop-filter bg-opacity-50'>
      <ul className='container flex items-center text-lg justify-around sm:justify-center px-2 sm:px-6 mx-auto capitalize dark:text-gray-300'>
        <li>
          <GoBack />
        </li>
        <li>
          <Search cb={handleSearch} />
        </li>

        <li>
          <NavLink
            end
            to='/'
            className={({ isActive }) =>
              textLinkClassName + (isActive ? activeTextLink : '')
            }
          >
            Recipes
          </NavLink>
        </li>

        <li>
          <NavLink
            to='favorites'
            className={({ isActive }) =>
              iconLinkClassName + (isActive ? activeIconLink : '')
            }
          >
            <BsFillBookmarkFill className='fill-current' />
          </NavLink>
        </li>
      </ul>
    </nav>
  );
};
