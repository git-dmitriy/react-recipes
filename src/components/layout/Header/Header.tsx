import { NavLink } from 'react-router-dom';
import { Search } from 'components/search/Search';
import { GoBack } from 'components/layout/GoBack';
import { BsFillBookmarkFill } from 'react-icons/bs';
import { ThemeSwitcher } from 'components/ThemeSwitcher';

export const Header: React.FC = () => {
  const textLinkClassName =
    'border-b-2 border-transparent font-bold ' +
    'outline-none hover:opacity-70 mx-1.5 sm:mx-6 transition-opacity';
  const iconLinkClassName =
    'block border-b-2 border-transparent outline-none hover:opacity-70 ' +
    'ml-1.5 mr-0.5 sm:ml-6 sm:mr-2 transition-opacity text-2xl';
  const activeTextLink = ' border-b-2 border-red-500';
  const activeIconLink = ' text-red-500';

  return (
    <nav className='bg-yellow-300 shadow dark:bg-gray-800 fixed top-0 w-full backdrop-blur-md backdrop-filter bg-opacity-50'>
      <ul className='container flex items-center text-lg justify-around sm:justify-center px-2 sm:px-6 mx-auto capitalize'>
        <li className='fixed top-28 left-4 sm:static'>
          <GoBack />
        </li>
        <li>
          <Search />
        </li>

        <li>
          <NavLink
            end
            to='/'
            className={({ isActive }) =>
              textLinkClassName + (isActive ? activeTextLink : '')
            }
          >
            Recipes
          </NavLink>
        </li>

        <li>
          <NavLink
            to='favorites'
            className={({ isActive }) =>
              iconLinkClassName + (isActive ? activeIconLink : '')
            }
          >
            <BsFillBookmarkFill className='fill-current' />
          </NavLink>
        </li>
        <li className='flex items-center'>
          <ThemeSwitcher />
        </li>
      </ul>
    </nav>
  );
};
