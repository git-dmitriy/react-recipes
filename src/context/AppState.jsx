import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';

export const AppState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
    theme: window.localStorage.getItem('theme') || 'light',
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  return (
    <AppContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        switchTheme,
        state,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
