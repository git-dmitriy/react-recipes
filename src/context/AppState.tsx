import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
  SET_IS_LOADING,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';
import { StateTypes, MealItemTypes, ContextTypes } from 'appTypes';

type P = { children: React.ReactNode };

export const AppState = ({ children }: P) => {
  const favorites = window.localStorage.getItem('favorites') || '[]';

  const initialState: StateTypes = {
    favorites: JSON.parse(favorites),
    theme: window.localStorage.getItem('theme') || 'light',
    isLoading: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item: MealItemTypes) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item: MealItemTypes) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode: string) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  const setIsLoading = (status = false) => {
    dispatch({
      type: SET_IS_LOADING,
      payload: status,
    });
  };

  const value: ContextTypes = {
    addToFavorites,
    removeFromFavorites,
    switchTheme,
    setIsLoading,
    state,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};
