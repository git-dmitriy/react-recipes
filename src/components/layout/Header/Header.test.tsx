import renderer from 'react-test-renderer';
import { Header } from 'components/layout/Header';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { AppState } from 'context/AppState';

it('renders header correctly', () => {
  const history = createMemoryHistory();

  const tree = renderer
    .create(
      <AppState>
        <Router location={history.location} navigator={history}>
          <Header />
        </Router>
      </AppState>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
