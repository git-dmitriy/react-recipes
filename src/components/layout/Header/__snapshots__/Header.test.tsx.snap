// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders header correctly 1`] = `
<nav
  className="bg-yellow-300 shadow dark:bg-gray-800 fixed top-0 w-full backdrop-blur-md backdrop-filter bg-opacity-50"
>
  <ul
    className="container flex items-center text-lg justify-around sm:justify-center px-2 sm:px-6 mx-auto capitalize"
  >
    <li
      className="fixed top-28 left-4 sm:static"
    >
      <div
        className="flex items-center sm:mx-6"
      >
        <button
          className="w-5 text-3xl hover:opacity-70 transition-opacity"
          onClick={[Function]}
        >
          <svg
            className="fill-current"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
            />
          </svg>
        </button>
      </div>
    </li>
    <li>
      <form
        className="flex"
        onSubmit={[Function]}
      >
        <div
          className="relative my-2 mx-auto dark:text-gray-900"
        >
          <input
            className="focus:outline-none focus:ring-2 dark:placeholder:text-gray-600 dark:bg-gray-100 focus:ring-red-500 shadow rounded-full w-52 sm:w-60 md:w-72 lg:w-80 xl:w-96 border-0 p-3 pl-9 min-w-full"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Search by name..."
            role="searchbox"
            type="text"
            value=""
          />
          <span
            className="absolute left-0 top-1/2 transform -translate-y-1/2 ml-4 "
            onClick={[Function]}
          >
            <svg
              className="fill-current"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
              />
            </svg>
          </span>
        </div>
      </form>
    </li>
    <li>
      <a
        aria-current="page"
        className="border-b-2 border-transparent font-bold outline-none hover:opacity-70 mx-1.5 sm:mx-6 transition-opacity border-b-2 border-red-500"
        href="/"
        onClick={[Function]}
      >
        Recipes
      </a>
    </li>
    <li>
      <a
        className="block border-b-2 border-transparent outline-none hover:opacity-70 ml-1.5 mr-0.5 sm:ml-6 sm:mr-2 transition-opacity text-2xl"
        href="/favorites"
        onClick={[Function]}
      >
        <svg
          className="fill-current"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 2v13.5a.5.5 0 0 0 .74.439L8 13.069l5.26 2.87A.5.5 0 0 0 14 15.5V2a2 2 0 0 0-2-2H4a2 2 0 0 0-2 2z"
          />
        </svg>
      </a>
    </li>
    <li
      className="flex items-center"
    >
      <button
        className="w-8 h-7 text-3xl relative overflow-hidden sm:hover:opacity-70 sm:transition-opacity"
        onClick={[Function]}
      >
        <span
          className=" absolute top-0 left-0 duration-200 transition-transform"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0a.996.996 0 000-1.41l-1.06-1.06zm1.06-10.96a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
            />
          </svg>
          <svg
            className="rotate-180"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 3a9 9 0 109 9c0-.46-.04-.92-.1-1.36a5.389 5.389 0 01-4.4 2.26 5.403 5.403 0 01-3.14-9.8c-.44-.06-.9-.1-1.36-.1z"
            />
          </svg>
        </span>
      </button>
    </li>
  </ul>
</nav>
`;
